Version 2

AlexWayfer opened the pull request:
Added
-----

* Built-in cache with file-path and max-age
* Automatically refresh from source (external repository)

Removed
-------

* Rake-tasks
* A very large copy of external UA-bots list
* `rack` as dependency

Changed
-------

* `voight_kampff/rack_request` isn't required by default
* Default path for cache is now in `./assets/`, not `./config/` (list of User-Agents is not config)

Development
-----------

* [EditorConfig](https://editorconfig.org/) added
* [RuboCop](https://github.com/rubocop-hq/rubocop) added and offenses fixed
* Drop support for Ruby <= 2.2 (they even will not get security patches)
* Ignore built gems for git

samuelgiles commented on lib/voight_kampff/test.rb:
```diff
@@ -1,6 +1,77 @@
+      end
+
+      def load_remote_crawlers(uri = DEFAULT_URL)
+        crawlers = Net::HTTP.get(URI(uri)).force_encoding('UTF-8')
```
Do you think this should potentially fail gracefully if say the URL returns a non successful response code or any kind of network error? Potentially just `Kernel.warn "VoightKampff: Unable to fetch crawlers from #{uri}"`

I can see a scenario where the repo gets pulled and a bunch of sites go down.

AlexWayfer replied:
Wow, thank you. I added tests and code for these situations.

Also, there is `warn` with error inspection: it can be `SocketError`, or `Errno::ENETUNREACH`, or something else.

And the code will "cache" the empty Array in these cases.

Also, I didn't add a code for redirects: `Net::HTTP` can't deal with them easily. It's more comfortable to use some external gem, like [`httpx`](https://gitlab.com/honeyryderchuck/httpx), for this. But I hope that there will be not redirects.

adamcrown replied:
Yes, if we did this at all, I'd want to be really paranoid about it. It should fail very gracefully and any sort of issues with this whole process should log an error, but never take down the whole site. And I'd prefer to avoid a situation where we need to give false positives too. If there are problems getting the file, I suppose you might need to treat everything as not a bot until the file did download. But that seems bad too.

At the very least, I think the gem should start off using it's local file thats packaged with the gem. Then only if a new version is downloaded successfully and the JSON is parsed successfully and then some sort of sanity check is done to make sure we have a real list of bots (even if that's just checking the number of bots in the file). Then and only then, should we start to use that new list.

adamcrown commented:
Wow, thanks for all the work on this @AlexWayfer! I'll do my best to get to reviewing it soon and see if we can get this merged. In the meantime, would you mind addressing the merge conflicts?

AlexWayfer commented:
> In the meantime, would you mind addressing the merge conflicts?

Done.

---

By the way, this version already works in production, and looks working properly.

AlexWayfer commented:
> I'll do my best to get to reviewing it soon and see if we can get this merged.

I don't want to rush, just a friendly reminder.

AlexWayfer commented:
Conflicts resolved, RuboCop updated, speed test improved.

adamcrown commented:
Sorry again for the delay. I've had a bit of a backlog with my open source projects ever since I started a job that gave me some time to work on them. But I'm trying to address at least one issue a week and this is currently next inline. So I should be able to get to it soon if nothing urgent comes up.

AlexWayfer commented:
> No external dependency, in general

I don't see evil in dependencies. Our (commercial) projects are depended on this gem. Is it bad? I don't think so.

Especially, I can contribute to dedicated repo only new bots, not a code, and this immediately will be in our projects, without updates. Like, automated data updating. Like currencies exchange rates.

> No downtime for the client app when GitHub or it's internet connection is down

It's a good point. This job (fetching data update) can be done in a separated thread and with retries. I can try to write a code for this here.

> No problems if the external file or repo is deleted, moved, broken, etc.

The same as first: our project can have problems if this gem will be deleted, moved, broken, etc. This is compromise, we're taking these risks. And these potential problems are resolvable.

> No performance issues from an external file having to be downloaded at startup and then on a regular interval

Ehm… OK. I think, it less than 10 seconds even with bad internet connection and large file, and for current interval (24 hours, that can be changed) it's like 0.01% of time, but OK. Again, with separated thread it can be better, I think.

> No issues with firewalls or running the project offline

Firewall with blocking downloads from GitHub?… And offline of projects with gem for requests?… I think, these cases are very rare, and there can be another problems. But OK, you're right.

> There certainly could be other cons I'm not thinking of too. Feel free to list them.

* Many commits in gem with code but about data from external repo, so it's kind of duplicating.

> I hope that's enough for you to see why I made the choice to not pull the file from GitHub by default.

Yes, of course.

> Let me know how you'd like to proceed with this.

I can (try to) separate those changes, but I want to describe my vision of suggested approach:

As referenced in the PR description, it's like [`money-oxr`](https://github.com/edlebert/money-oxr) for me. Gem for currencies exchange rates don't store a copy of all currency rates, which is (automatically) updating with interval. It's like… pointless. It (almost) requires regular updates from gem side, and duplicating a large massive of data. While this data is updating every day (or even maybe hour). 

The data for this gem (crawlers list) is not too frequently updating, but for me the principle the same: separate code and data. Like you separate code from database, code from uploaded files, etc. You don't store them in repo with code and update regularly.

AlexWayfer commented on .ruby-version:
```diff
@@ -1 +0,0 @@
-2.6.2
```
The gem supports many version of Ruby. Why we should lock at some specific version? It's like `Gemfile.lock` (https://stackoverflow.com/a/4151540/2630849), but Ruby-level.
